dagshub/data_engine/model/datasource.py

```diff
@@ -543,49 +617,111 @@ def _upload_metadata(self, metadata_entries: List[DatapointMetadataUpdateEntry])
         Save the dataset, which is a combination of datasource + query, on the backend.
         That way you can persist and share your queries.
         You can get the dataset back by calling :func:`.datasets.get_dataset()`
 
         Args:
             name: Name of the dataset
+            assign_after_save: Keep the reference to the dataset after saving.
+                This makes other functions like :func:`clear_query`, :func:`log_to_mlflow` and :func:`save_to_file`
+                refer to the dataset.
         """
         send_analytics_event("Client_DataEngine_QuerySaved", repo=self.source.repoApi)
 
         self.source.client.save_dataset(self, name)
         log_message(f"Dataset {name} saved")
 
+        if assign_after_save:
+            self.load_from_dataset(dataset_name=name, change_query=False)
+
+    def log_to_mlflow(self, artifact_name=DEFAULT_MLFLOW_ARTIFACT_NAME):
+        """
+        Logs the current datasource state to MLflow as an artifact.
+
+        Args:
+            artifact_name: Name of the artifact that will be stored in the MLflow run
+        """
+        run = mlflow.active_run()
+        mlflow.log_dict(self._to_dict(), artifact_name)
+        log_message(f'Saved the datasource state to MLflow (run "{run.info.run_name}") as "{artifact_name}"')
+
     def save_to_file(self, path: str = ".", name: str = "") -> str:
         """
-        [EXPERIMENTAL]
-        Saves a text file representing the datasource or dataset information to a file which can be committed to Git.
+        Saves a text file representing the current state of datasource or dataset
+        to a file ``<name>.json`` which can be committed to Git.
         Useful for connecting code versions to the datasource used for training.
 
         .. note::
-            Does not save dataset contents, only a pointer file.
+            Does not save dataset contents.
 
         Args:
-            path: Path to save the datasource or dataset file in, defaults to current directory
-            name: Optional: name of the file. if not provided, datasource or dataset name will be used instead.
+            path: Path to the directory where to save the datasource or dataset file.
+                Defaults to current working directory.
+            name: Optional: name of the file. if not provided, datasource name will be used instead.
 
         Returns:
             The path to the saved file
         """
-        res = DatasourceFileStruct(
-            id=self._source.id,
-            repo=self._source.repo,
-            name=self._source.name,
-            datasetQuery=self.serialize_gql_query_input()
-        )
-        file_path = os.path.join(path, (name or res.name) + ".dagshub")
+        file_path = os.path.join(path, (name or self.source.name) + ".json")
+        res = self._to_dict()
         with open(file_path, "w") as file:
-            file.write(json.dumps(res.to_dict(), indent=4))
+            file.write(json.dumps(res, indent=4))
```

Comment:
I'm pretty sure you can write directly to the file + sort by keys maybe, for consistent ordering?
Or else, use an ordered dict.
Or does your dataclass output ordered dicts by default?
```suggestion
            json.dump(res, file, indent=4, sort_keys=True)
```

Reply:
good idea, will change that